File: samples/face_alignment/hih/hih.py
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)

import cv2
import argparse
from hih_aligner.hih_aligner import Hih
import common.utils as utils


def argument_parser():
    parser = argparse.ArgumentParser(description="FACE_ALIGNMENT")
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/hih_2s-int8-percentile-1_3_256_256-vacc/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--vdsp_params",
        default="../data/configs/hih_bgr888.json",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="../../../../data/images/face.jpg",
        type=str,
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="./face_out.png",
        type=str,
        help="output file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        type=str,
        help="input dataset image list",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="dataset root",
    )
    parser.add_argument(
        "--dataset_output_file",
        default="",
        type=str,
        help="dataset output file",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    face_aligner = Hih(args.model_prefix, args.vdsp_params, batch_size, args.device_id)
    image_format = face_aligner.get_fusion_op_iimage_format()
    text = []
    if args.dataset_filelist == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Read image failed:{args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)
        landmarks = face_aligner.process(vsx_image)
        print("Face alignment results:")
        print(landmarks)
    else:
        filelist = []
        with open(args.dataset_filelist, "rt") as f:
            filelist = [line.strip() for line in f.readlines()]
        for _, file in enumerate(filelist):
            filename = os.path.basename(file)
            fullname = os.path.join(args.dataset_root, filename)
            cv_image = cv2.imread(fullname)
            assert cv_image is not None, f"Read image failed:{fullname}"
            vsx_image = utils.cv_bgr888_to_vsximage(
                cv_image, image_format, args.device_id
            )
            landmarks = face_aligner.process(vsx_image)
            text.append(filename)
            for i in landmarks.reshape((1, -1)).squeeze():
                text.append(f" {i}")
            text.append("\n")

    if args.dataset_output_file != "":
        with open(args.dataset_output_file, "w+") as f:
            f.writelines(text)

File: samples/face_alignment/hih/hih.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <fstream>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "hih_aligner/hih_aligner.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("model_prefix", 'm',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "hih_2s-int8-percentile-1_3_256_256-vacc/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/hih_bgr888.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input file", false,
                        "../data/images/face.jpg");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_file", '\0', "dataset output file",
                        false, "");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  const int batch_size = 1;
  vsx::Hih face_aligner(args.get<std::string>("model_prefix"),
                        args.get<std::string>("vdsp_params"), batch_size,
                        args.get<uint32_t>("device_id"));
  auto image_format = face_aligner.GetFusionOpIimageFormat();

  if (args.get<std::string>("dataset_filelist").empty()) {
    auto cv_image = cv::imread(args.get<std::string>("input_file"));
    CHECK(!cv_image.empty())
        << "Failed to read image:" << args.get<std::string>("input_file")
        << std::endl;
    vsx::Image vsx_image;
    CHECK(vsx::MakeVsxImage(cv_image, vsx_image, image_format) == 0);
    auto result = face_aligner.Process(vsx_image);
    const float* array_data = result.Data<float>();
    std::cout << "Face alignment results: \n";
    for (size_t i = 0; i < result.GetSize(); i += 2) {
      std::cout << "(" << array_data[i] << ", " << array_data[i + 1] << ")"
                << std::endl;
    }
  } else {
    std::vector<std::string> filelist =
        vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
    auto dataset_root = args.get<std::string>("dataset_root");
    std::ofstream outfile(args.get<std::string>("dataset_output_file"));
    CHECK(outfile.is_open())
        << "Failed to open: " << args.get<std::string>("dataset_output_file");
    for (auto file : filelist) {
      auto fullname = file;
      if (!dataset_root.empty()) fullname = dataset_root + "/" + file;
      auto cv_image = cv::imread(fullname);
      CHECK(!cv_image.empty())
          << "Failed to read image:" << fullname << std::endl;
      vsx::Image vsx_image;
      CHECK(vsx::MakeVsxImage(cv_image, vsx_image, image_format) == 0);
      auto result = face_aligner.Process(vsx_image);
      const float* array_data = result.Data<float>();
      outfile << file << " ";
      for (size_t i = 0; i < result.GetSize(); i++) {
        outfile << array_data[i] << " ";
      }
      outfile << std::endl;
    }
    outfile.close();
  }

  return 0;
}